### ISSUE TITLE
`unnecessary_unwrap`: false positive with disjunctive `if`-condition

### ISSUE THREAD

**Issue by tsdh** [rust-lang/rust-clippy#8535]
### Summary

I think I get a false positive when having an `if-else` where the condition is a disjunction where the tested `Option` may be `None` or have some specific value.

### Lint Name

unnecessary_unwrap

### Reproducer

I tried this code:

```rust
fn main() {
    let opt: Option<i32> = None;
    if opt.is_none() || opt.unwrap().is_positive() {
        println!("None or positive")
    } else {
        foo(opt.unwrap())
    }
}

fn foo(x: i32) {
    println!("foo: {}", x)
}
```

I saw this happen:

```
/home/horn/.cargo/bin/cargo clippy --manifest-path /home/horn/tmp/playground-rs/Cargo.toml 
    Checking playground-rs v0.1.0 (/home/horn/tmp/playground-rs)
warning: called `unwrap` on `opt` after checking its variant with `is_none`
 --> src/main.rs:6:13
  |
3 |     if opt.is_none() || opt.unwrap().is_positive() {
  |        ------------- the check is happening here
...
6 |         foo(opt.unwrap())
  |             ^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::unnecessary_unwrap)]` on by default
  = help: try using `if let` or `match`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap

warning: `playground-rs` (bin "playground-rs") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s

Cargo-Process finished at Mon Mar 14 15:56:21
```

I expected to see this happen:

There should be no warning as there is no good way to rewrite this code using `if let` or `match`.  Feel free to prove me wrong, though. :-)

### Version

```text
rustc 1.59.0 (9d1b2106e 2022-02-23)
binary: rustc
commit-hash: 9d1b2106e23b1abd32fce1f17267604a5102f57a
commit-date: 2022-02-23
host: x86_64-unknown-linux-gnu
release: 1.59.0
LLVM version: 13.0.0

$ cargo clippy --version
clippy 0.1.59 (9d1b210 2022-02-23)
```

**Comment by Jarcho** [comment]
You could do:
```rust
match opt {
    Some(x) if !x.is_positive() => { ... },
    _ => { ... }
}
```

**Comment by tsdh** [comment]
True, although I'm not convinced that this is a big improvement...

**Comment by xtqqczze** [comment]
Related: https://github.com/rust-lang/rust-clippy/issues/15744#issuecomment-3690564277
